### PR TITLE
Update layer and service managers to support new approach

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

**Review by Eabramov84 — APPROVED** [review]

File: servicemanager/servicemanager.go [hunk 1]
```diff
@@ -211,101 +182,46 @@ func (sm *ServiceManager) Close() {
...
-}
-
-// UseService sets last use of service.
-func (sm *ServiceManager) UseService(serviceID string, aosVersion uint64) error {
-	if err := sm.serviceInfoProvider.SetServiceTimestamp(serviceID, aosVersion, time.Now().UTC()); err != nil {
-		return aoserrors.Wrap(err)
-	}
-
-	serviceInfo, err := sm.serviceInfoProvider.GetService(serviceID)
+// ProcessDesiredLayers installs, removes, restores desired services on the system.
```
Comment (Eabramov84) [inline servicemanager/servicemanager.go:206]: // incorrect comment should be ProcesdDesiredServices

Comment (MykolaSolyanko) [reply servicemanager/servicemanager.go:206]: Done

File: layermanager/layermanager.go [hunk 2]
```diff
@@ -196,110 +191,97 @@ func (layermanager *LayerManager) Close() {
...
+		}
 	}
 
-	return nil
+	return desiredLayers, nil
 }
 
-func (layermanager *LayerManager) doInstallLayer(
-	installInfo LayerInfo, layerURL string, fileInfo image.FileInfo,
+/***********************************************************************************************************************
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply layermanager/layermanager.go:264]: Done

File: layermanager/layermanager.go [hunk 3]
```diff
@@ -196,110 +191,97 @@ func (layermanager *LayerManager) Close() {
...
 
-		return nil
-	}
+func (layermanager *LayerManager) cachedLayers(
+	desiredLayers []aostypes.LayerInfo, storeLayers []LayerInfo,
+) ([]aostypes.LayerInfo, error) {
+nextLayer:
+	for _, storeLayer := range storeLayers {
+		for i, desiredLayer := range desiredLayers {
+			if desiredLayer.Digest == storeLayer.Digest {
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply layermanager/layermanager.go:241]: Done

File: layermanager/layermanager.go [hunk 4]
```diff
@@ -196,110 +191,97 @@ func (layermanager *LayerManager) Close() {
...
 	}
 
-	if !layer.Cached {
-		log.Warningf("Layer %v not cached", digest)
+	return nil
+}
 
-		return nil
-	}
+func (layermanager *LayerManager) cachedLayers(
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply layermanager/layermanager.go:235]: Done

File: servicemanager/servicemanager.go [hunk 5]
```diff
@@ -163,34 +144,23 @@ func New(
...
-			size, err := sm.getServiceSize(service.ServiceID)
-			if err != nil {
-				return nil, aoserrors.Wrap(err)
-			}
-
-			if err = sm.serviceAllocator.AddOutdatedItem(
-				service.ServiceID, size, service.Timestamp); err != nil {
-				return nil, aoserrors.Wrap(err)
-			}
+		if err := sm.serviceAllocator.AddOutdatedItem(
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply servicemanager/servicemanager.go:153]: Done.

File: servicemanager/servicemanager.go [hunk 6]
```diff
@@ -323,60 +239,66 @@ func (sm *ServiceManager) ValidateService(service ServiceInfo) error {
  * Private
  **********************************************************************************************************************/
 
-func (sm *ServiceManager) doRestoreService(serviceID string) error {
-	log.WithFields(log.Fields{"serviceID": serviceID}).Debug("Restore service")
+func (sm *ServiceManager) cachedService(
```
[the remark this answers is not in the document]
Comment (MykolaSolyanko) [reply servicemanager/servicemanager.go:242]: Done

File: servicemanager/servicemanager.go [hunk 7]
```diff
@@ -163,24 +144,15 @@ func New(
...
-
 		if service.Cached {
-			size, err := sm.getServiceSize(service.ServiceID)
-			if err != nil {
-				return nil, aoserrors.Wrap(err)
-			}
-
-			if err = sm.serviceAllocator.AddOutdatedItem(
-				service.ServiceID, size, service.Timestamp); err != nil {
+			if err := sm.serviceAllocator.AddOutdatedItem(
```
Comment (al1img) [inline servicemanager/servicemanager.go:154]: Looks like long line (more than 120).
if !cached -> continue? To decrease nested levels.

**Review by al1img — APPROVED** [review]

File: servicemanager/servicemanager.go (hunk 7 above)
Comment (MykolaSolyanko) [reply servicemanager/servicemanager.go:154]: Done. 
line less than 120 characters

**Review by al1img — APPROVED** [review]

**Pull request merged**